File: 2020/10/main.hs
import Data.List (sort)

answer1 :: [Int] -> (Int, Int) -> Int
answer1 [_] (a, b) = a * b
answer1 (x : xs) (a, b)
  | diff == 1 = answer1 xs (a + 1, b)
  | diff == 3 = answer1 xs (a, b + 1)
  where
    diff = head xs - x

answer2 :: [Int] -> Int -> Int
answer2 _ 0 = 1
answer2 points point
  | point `notElem` points = 0
  | otherwise = childPaths
  where
    childPaths = sum [answer2 points (point - x) | x <- [1 .. 3]]

main :: IO ()
main = do
  lines <- fmap lines (readFile "input.txt")
  let numbers = sort $ map read lines :: [Int]
  let chargers = [0] ++ numbers ++ [maximum numbers + 3]
  print $ answer1 chargers (0, 0)
  print $ answer2 chargers (maximum chargers)

File: 2020/10/main.py
with open("input.txt") as f:
    lines = f.readlines()

lines = sorted([int(x) for x in lines])

lines = [0] + lines + [(max(lines) + 3)]
one_diffs = 0
three_diffs = 0

for i, x in enumerate(lines):
    if i + 1 == len(lines):
        break

    diff = lines[i + 1] - x
    if diff == 3:
        three_diffs += 1
    elif diff == 1:
        one_diffs += 1

answer1 = one_diffs * three_diffs
print(answer1)

all_paths = {}


def route(point, points):
    if point in all_paths:
        return all_paths[point]

    paths = 0
    for n in range(1, 4):
        x = point - n
        if x == 0:
            paths += 1
        elif x not in points:
            continue
        else:
            paths += route(x, points)

    all_paths[point] = paths
    return paths


answer2 = route(lines[-1], lines)

print(answer2)
